File: algorithms/[150]逆波兰表达式求值/solution.py
# topics = ["栈"]

import re
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack: List[int] = []

        for ch in tokens:
            if re.match(r'^-?\d+$', ch):
                stack.append(int(ch))
            elif ch == '+':
                stack.append(stack.pop() + stack.pop())
            elif ch == '*':
                stack.append(stack.pop() * stack.pop())
            elif ch == '-':
                b = stack.pop()
                a = stack.pop()
                stack.append(a - b)
            elif ch == '/':
                b = stack.pop()
                a = stack.pop()
                # 浮点数取整
                stack.append(int(a / b))

        return stack[-1]

File: algorithms/[150]逆波兰表达式求值/solution.ts
// topics = ["栈"]

function evalRPN(tokens: string[]): number {
  const stack: number[] = [];

  for (let i = 0, n = tokens.length; i < n; i += 1) {
    const ch = tokens[i]!;

    if (/[0-9]/.test(ch)) {
      stack.push(Number(ch));
    } else if (ch == '+') {
      stack.push(stack.pop()! + stack.pop()!);
    } else if (ch == '*') {
      stack.push(stack.pop()! * stack.pop()!);
    } else if (ch == '-') {
      const b = stack.pop()!;
      const a = stack.pop()!;
      stack.push(a - b);
    } else if (ch == '/') {
      const b = stack.pop()!;
      const a = stack.pop()!;
      // 位运算取整
      stack.push(~~(a / b));
    }
  }

  return stack[stack.length - 1];
}
